### PR TITLE
Improve unsafe hygiene

### PR DESCRIPTION
**Pull request by Cassy343**
Should contain all of, and only the desired changes discussed in https://github.com/faern/oneshot/pull/17. If there are any additional changes desired, I'll do them on this branch and deal with merge/rebase conflicts after this is merged.

### CODE REVIEW

**Review by faern — COMMENTED** [review]
Awesome! I just glanced over this quickly as I have to go out. I have read most of the code already, so this looks pretty good already! I'll look at the code in more detail in a day or two.

File: src/lib.rs [hunk 1]
```diff
@@ -682,53 +724,114 @@ use states::*;
...
+        #[cfg(not(loom))]
+        {
+            op(&mut *self.message.get())
+        }
     }
 
     #[inline(always)]
-    unsafe fn drop_message(&mut self) {
-        ptr::drop_in_place(self.message.as_mut_ptr());
+    unsafe fn with_waker_mut<F>(&self, op: F)
```
[the remark this answers is not in the document]
Comment (faern) [reply src/lib.rs:770]: Personally I'm not in the MSRV-change-is-a-breaking-change camp. I'd much rather just use a newer rustc than needing these workarounds. But either way works. But we can't allow compiler warnings no matter the combinations of features used to compile.

File: src/errors.rs [hunk 2]
```diff
@@ -1,50 +1,64 @@
...
 use core::mem;
+use core::ptr::NonNull;
 
 /// An error returned when trying to send on a closed channel. Returned from
 /// [`Sender::send`] if the corresponding [`Receiver`] has already been dropped.
 ///
 /// The message that could not be sent can be retreived again with [`SendError::into_inner`].
 pub struct SendError<T> {
-    channel_ptr: *mut super::Channel<T>,
+    channel_ptr: NonNull<Channel<T>>,
```
[the remark this answers is not in the document]
Comment (faern) [reply src/errors.rs:11]: Since it should preferably be cheap to sent and ignore the error. Not having to pay for that if you don't use the error for anything

[the remark this answers is not in the document]
Comment (Cassy343) [reply src/errors.rs:11]: That makes sense. One thing to note is that the *correct* way to fully implement box involves using the unstable `dropck_eyepatch` feature to add `#[may_dangle]` to the `Drop` implementation, but all that does it make it slightly *less* restrictive which is better than the alternative.

TL;DR we should add a `PhantomData<T>` field, which I'll do later today.

[the remark this answers is not in the document]
Comment (faern) [reply src/errors.rs:11]: Great! To add to my last point: SendError should be small, so that the stack size to call Sender::send stays small (for small T).

[the remark this answers is not in the document]
Comment (Cassy343) [reply src/errors.rs:11]: Yeah that's a good point as well, and `NonNull` helps with that too since `Result<(), SendError>` will decrease from 16 bytes to 8 (on 64-bit architectures), so it can fit in a register.

File: src/lib.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (Cassy343) [reply src/lib.rs:770]: I think we can just hold off on bumping the MSRV for this until `IntoFuture` lands in 1.64, then we batch this with the impl on `Receiver`. I'll work on fixing the compiler warnings.

File: src/errors.rs [hunk 3]
```diff
@@ -1,50 +1,69 @@
...
+    _dropck: PhantomData<T>,
```
Comment (faern) [inline src/errors.rs:13]: Great! But I think we need an explanation for this. And a link pointing to further info. It's very unclear why this is here just from looking at this code alone. I think we should mention the Box struct and maybe have the same link they have. M

Also, do you have a good understanding of what difference this actually makes? Code seems to compile and work without it. I don't have a good intuition yet.

Comment (Cassy343) [reply src/errors.rs:13]: I don't recall the exact reason off the top of my head - it's highly technical and has to do with the fact that types can run arbitrary code when they're dropped, and that because of that in order for a generic type to be able to be dropped soundly it must strictly outlive its container. There's more information [here](https://doc.rust-lang.org/nomicon/dropck.html).

Comment (Cassy343) [reply src/errors.rs:13]: > Code seems to compile and work without it.

It compiles and works without it because without the phantom data *too much* code is allowed to compile, including code which exhibits undefined behavior, which is why it's needed.

Comment (faern) [reply src/errors.rs:16]: > It compiles and works without it because without the phantom data too much code is allowed to compile, including code which exhibits undefined behavior, which is why it's needed.

Since this is not obvious, the best would be if we could test for this. Do you have an example of unsound code that would compile without the `PhantomData` but not with it? If so, we should consider adding it as a compile-fail test. Maybe with the help of [`compiletest_rs`](https://crates.io/crates/compiletest_rs). I do understand the basics of why this is needed to be sound. But I also can't come up with any actual code that would trigger UB. Since the `T` can't possibly reference data in neither `Channel` nor `SendError`/`Sender`/`Receiver`.

Also.. Since I guess the old `*mut super::Channel<T>` code was letting unsound code compile also, then this change is fixing an old bug, not just making the new design sound.

File: src/errors.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (faern) [reply src/errors.rs:12]: This is a great improvement! Consider adding this as a test, so we make sure the null-pointer optimization does not have a regression further down the line. We can add this to the existing `error_sized` test:

```rust
    // Makes sure that the return value of `Sender::send` stays small. In essence
    // checks that the null-pointer optimization in `SendError` works.
    assert_eq!(
        mem::size_of::<Result<(), oneshot::SendError<[u8; 1024]>>>(),
        EXPECTED
    );
```

File: src/errors.rs [hunk 4]
```diff
@@ -1,50 +1,72 @@
...
 }
 
 unsafe impl<T: Send> Send for SendError<T> {}
+unsafe impl<T: Sync> Sync for SendError<T> {}
 
 impl<T> SendError<T> {
-    pub(crate) const fn new(channel_ptr: *mut super::Channel<T>) -> Self {
-        Self { channel_ptr }
+    /// Safety: by calling this function, the caller semantically transfers ownership of the
```
Comment (faern) [inline src/errors.rs:23]: Nitpicky. But this should be in the format `/// # Safety`. Since it's documentation on how the function should be used.

File: src/lib.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (faern) [reply src/lib.rs:771]: The last release is `0.1.3`. That release had an MSRV of 1.51. Commit 90f22768c26e1c90b972cdc5669bed4a7c876a24 already bumped the MSRV for the upcoming release. I see no benefit of bumping it from 1.51 -> 1.56 over 1.51 -> 1.60, which would be the case if you used the nice helpers stabilized in 1.60 here.

**Review by faern — CHANGES_REQUESTED** [review]
Looks good! I have some comments and feedback to be addressed/discussed. But overall, it looks almost mergeable.

File: src/errors.rs (hunk 3 above)
Comment (Cassy343) [reply src/errors.rs:16]: Without the phantom data, this code would trigger UB:
```rs
fn bad() {
    let (tx, rx) = oneshot::channel::<Box<u8>>();
    drop(rx);
    let error = tx.send(Box::new(0)).unwrap_err();

    struct Oof<'a>(&'a u8);

    impl<'a> Drop for Oof<'a> {
        fn drop(&mut self) {
            println!("{}", self.0);
        }
    }

    struct Foo<'a> {
        error: SendError<Box<u8>>,
        oof: Option<Oof<'a>>,
    }

    let mut foo = Foo {
        error,
        oof: None
    };
    foo.oof = Some(Oof(&**foo.error.as_inner()));
    drop(foo);
}```
Since fields are dropped in order of declaration, `error` gets dropped, freeing the `Box<u8>`, and then `Oof` gets dropped, accessing the data behind the box, which is a use-after-free. With the phantom data this code fails to compile.

Comment (Cassy343) [reply src/errors.rs:16]: > However, after reading the docs a bit more carefully I think we need these `_dropck` markers on `Sender` and `Receiver` also. Right?

I don't believe so. This issue only shows up when you allow someone to obtain a reference to the data you're holding, and since `Sender` and `Receiver` only take and give ownership of the message respectively, we won't run into that problem.

That being said, you are correct that either *could* drop a `T`, so I'm not opposed to adding that phantom data in case I'm missing something, since it's not incorrect.

File: src/errors.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (Cassy343) [reply src/errors.rs:12]: Ah yeah I did add a test for that in one of the later commits from the draft PR. I'll move those over and deal with the rebase conflicts later cause it's a small easy change.

File: src/errors.rs (hunk 3 above)
Comment (Cassy343) [reply src/errors.rs:16]: > If so, we should consider adding it as a compile-fail test. Maybe with the help of [`compiletest_rs`](https://crates.io/crates/compiletest_rs).

I added a test for that, but we actually don't have to use `compiletest_rs`. If you add docs to private items, and add tests in those docs, then those doc tests will still be run, so I just added a compile fail test on `_dropck` to test that.

File: src/lib.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (Cassy343) [reply src/lib.rs:779]: Sounds good. Bumped the MSRV and switched back to the utility method.

File: src/errors.rs (hunk 3 above)
Comment (faern) [reply src/errors.rs:47]: Nicely done! Having the documentation test directly on the field makes it even better.

**Comment by faern** [comment]
Awesome! I think everything is addressed now and ready for merge. I'll have another look at the code, and hopefully merge a bit later today.

**Review by faern — APPROVED** [review]

**Comment by faern** [comment]
@Cassy343 Do you want to clean up/rewrite the git history? Otherwise I'll merge this after I've had some food :)

**Comment by Cassy343** [comment]
You can go ahead and squash and merge, I always worry I'll mess something up when I edit git history so I'd rather not haha.

**Comment by faern** [comment]
Thanks for the contribution! Please come with feedback on the changelog entries I added for this: https://github.com/faern/oneshot/commit/3b27d35104845d9c76fabd6be1859c56377b57f3.

Looking forward to the remaining fixes and improvements :)

**Comment by Cassy343** [comment]
The changelog entries look good to me!

File: src/errors.rs (hunk 3 above)
Comment (faern) [reply src/errors.rs:47]: @Cassy343 Were you able to verify that this test actually failed without the `_dropck` field? Because I'm trying it now, but I can not make the code exhibit the UB that the test is supposed to catch. Yes, the test does not compile, but for a very different reason:

```
error[E0412]: cannot find type `SendError` in this scope
  --> src/errors.rs:34:12
   |
17 |     error: SendError<Box<u8>>,
   |            ^^^^^^^^^ not found in this scope
   |
help: consider importing one of these items
   |
2  | use async_std::channel::SendError;
   |
2  | use oneshot::SendError;
   |
2  | use std::sync::mpsc::SendError;
```

**Comment by faern** [comment]
Please have a look at this comment: https://github.com/faern/oneshot/pull/18/files#r922725443. Am I missing something here, or is the test not testing what it's supposed to test?

File: src/errors.rs (hunk 3 above)
Comment (Cassy343) [reply src/errors.rs:47]: Oh I verified the test worked in an actual function locally before moving it to the doc test, just import the `SendError` and it should fail for the correct reason

Comment (faern) [reply src/errors.rs:47]: No, for me it fails with the exact same error with and without the `PhantomData<T>`:

```
error[E0597]: `foo.error` does not live long enough
  --> src/errors.rs:44:23
   |
26 | foo.oof = Some(Oof(&**foo.error.as_inner()));
   |                       ^^^^^^^^^^^^^^^^^^^^ borrowed value does not live long enough
27 | drop(foo);
28 | } _doctest_main_src_errors_rs_21_0() }
   | -
   | |
   | `foo.error` dropped here while still borrowed
   | borrow might be used here, when `foo` is dropped and runs the destructor for type `Foo<'_>`

error[E0505]: cannot move out of `foo` because it is borrowed
  --> src/errors.rs:45:6
   |
26 | foo.oof = Some(Oof(&**foo.error.as_inner()));
   |                       -------------------- borrow of `foo.error` occurs here
27 | drop(foo);
   |      ^^^
   |      |
   |      move out of `foo` occurs here
   |      borrow later used here
```

Comment (Cassy343) [reply src/errors.rs:47]: Okay I got some stuff mixed up, it turns out we *don't* need `PhantomData<T>` because we're not using `#[may_dangle]` (the nomicon wasn't clear about this and I misremembered). That compile fail test can be removed since it will fail either way, however the phantom data should still be left there in case the drop check becomes more advanced, or a `#[may_dangle]`-like feature becomes stabilized or the compiler can assume when a type may dangle.

[This video](https://www.youtube.com/watch?v=TJOFSMpJdzg&t=1612s) explains it best (just watching the "Boks is not restrictive enough" section should be sufficient).
